### PR TITLE
ls: Improve error handling -- Match GNU semantics and ordering

### PR DESCRIPTION
**Pull request by kimono-koans**
This PR improves ls error handling by better matching the GNU output and ordering for many common ls errors.

Referenced here as well: https://github.com/uutils/coreutils/issues/2785

### CODE REVIEW

File: src/uu/ls/src/ls.rs
Comment (sylvestre) [inline src/uu/ls/src/ls.rs:4]: please remove this, it isn't adding much value + you will be credited in the git history

File: src/uu/ls/src/ls.rs
Comment (sylvestre) [inline src/uu/ls/src/ls.rs:2058]: please fix the clippy warning

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:4]: Yeah, no problem.

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:2057]: Sure, done.

File: src/uu/ls/src/ls.rs
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:1606]: Also, what was the problem exactly with ownership and flushing? The following code works just fine:
```rust
use std::io::{stdout, BufWriter, Write};

fn main() {
    let mut w = BufWriter::new(stdout());

    let _ = vec![1, 2, 3, 4].iter().filter_map(|&x| {
        let _ = w.flush();
        if x < 3 {
            Some(x)
        } else {
            None
        }
    });
}
```

And it also works with another function:
```rust
use std::io::{stdout, BufWriter, Write, Stdout};

fn main() {
    let mut w = BufWriter::new(stdout());

    let _ = vec![1, 2, 3, 4].iter().filter_map(|&x| {
        do_something(x, &mut w)
    });
}

fn do_something(x: i32, out: &mut BufWriter<Stdout>) -> Option<i32> {
    let _ = out.flush();
    if x < 3 {
        Some(x)
    } else {
        None
    }
}
```

Is there something here that makes this impossible?

File: src/uu/ls/src/ls.rs
Comment (tertsdiepraam) [inline src/uu/ls/src/ls.rs:2076]: Please run `cargo fmt`

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Thanks! Flushing was a great idea! I left a few suggestions, the biggest of which is to use `UIoError`, for which you can find the documentation [here](https://docs.rs/uucore/0.0.10/uucore/error/struct.UIoError.html).

File: src/uu/ls/src/ls.rs
Comment (tertsdiepraam) [inline src/uu/ls/src/ls.rs:1243]: Is this leftover from some experimentation?

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:145]: Thank you for your review.  I popped an error in the test after naively trying UIoError.  

> thread 'test_ls::test_ls_files_dirs' panicked at ''ls: cannot access 'doesntexist': Entity not found ' does not contain ''doesntexist': No such file or directory'', tests/common/util.rs:396:9

"Entity not found" does not match GNU error codes.  I think it should be "No such file or directory." I'm more than pleased to fix myself but do you have a utility which you would consider uses best practices, so I can implement according to the preferred approach?

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:2079]: Agreed this is much better looking.

File: tests/by-util/test_ls.rs
Comment (kimono-koans) [reply tests/by-util/test_ls.rs:2336]: Yes, that makes sense.

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:1606]: That looks much more idiomatic!  Thanks.

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:2076]: I think `cargo fmt` is what is doing this!  I run rust fmt and cargo fmt before each commit.

File: src/uu/ls/src/ls.rs
Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:1243]: Yes, caught me!  Thank you.

File: src/uu/ls/src/ls.rs
Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:146]: So, what I would do is to make this variant `IOError(UIoError)` and implement `From<UIoError>` for `LsError`. Something like this:
```rust
let unwrapped = match entry {
    Ok(path) => path,
    Err(error) => {
        let _ = out.flush();
        show!(LsError::IOError(error));
        break;
    }
};
```
then becomes something like this:
```rust
use uucore::{error::FromIo, display::Quotable};
...
let unwrapped = entry.map_err_context(|| name.maybe_quote().to_string())?;
```

Surprisingly, there's no util that does exactly this. But [`install`](https://github.com/uutils/coreutils/blob/master/src/uu/install/src/install.rs) has a slightly different approach, which is also quite nice.

Comment (tertsdiepraam) [reply src/uu/ls/src/ls.rs:146]: The "Entity not found" error is interesting. If you run into it again, please show me how it happens, because that would be a bug. UIoError should already show the "No such file or directory" message instead of "Entity not found"

Comment (kimono-koans) [reply src/uu/ls/src/ls.rs:146]: I think you may have a bug then.  I just pushed my naive implementation.  Let me see if I can point you to it.

**Review by kimono-koans — COMMENTED** [review]
See my implementation here.

**Comment by tertsdiepraam** [comment]
You're right that there is a bug! I'll check it out.

**Comment by kimono-koans** [comment]
> You're right that there is a bug! I'll check it out.

Now I'm not so sure...  "Entity not found" could be output from the test harness?  Let me kick it around too.

**Comment by tertsdiepraam** [comment]
So the problem is that you're constructing the error from scratch, instead of basing it on an OS error, like opening a file would do. The implementation of the message only takes those with raw OS errors into account. The breaking change was introduced in [this PR](https://github.com/uutils/coreutils/pull/2675/files). I'll try to think of a solution.

**Comment by tertsdiepraam** [comment]
Okay, here's my proposal. The situation for `ls` is that we access we metadata in a lot of places. So this is stored in the `PathData` struct with a `OnceCell`. The idea is that the metadata is only retrieved from the OS once and that further accesses will reuse the same information. When I made this change, however, I did not think about what that means for error handling. Currently, we print the errors at the callsites of `PathData::md` (and by extension `PathData::file_type`), but what we should do instead, is print the errors inside of `PathData::md`. This ensures both that we have access to the actual `io::Error` and also that the error message is printed only once. So, `md` and `file_type` should take a mutable reference to `out` and print an error to it if the metadata cannot be retrieved.

If you want (and agree with this proposal), you can try to make that change here. Otherwise, I could try once this is merged.

**Comment by kimono-koans** [comment]
> If you want (and agree with this proposal), you can try to make that change here. Otherwise, I could try once this is merged.

As I understand your proposal -- ask for metadata once, store the Result in the PathData struct, and when called, if there is any error, display the error.  This approach seems more efficient, but I think it may complicate matters before we nail down the actual GNU semantics. 

My modest opinion is -- I think the best thing to do is learn a little more first.  I think we use my/our current custom io::Error wrapper (which we think it approximately matches the GNU semantics), and write a bunch of tests (I still haven't piped stderr to stdout in a test to check for order) that nail down how GNU actually prints errors for ls, then merge, and then take a look at what might be done better/how to use UIoError, so that when we test a new approach we know more about how it should work in practice.

What do you think?

**Comment by tertsdiepraam** [comment]
Alright! I'll pick this up after this is merged. I think we can merge this if you clean the experimentation up a bit and fix fmt and lint issues.

**Comment by kimono-koans** [comment]
> Alright! I'll pick this up after this is merged. I think we can merge this if you clean the experimentation up a bit and fix fmt and lint issues.

The more I've tooled around with this -- I've come to the conclusion -- you were right.  I'm going to take a stab at implementing.  Again, really appreciated and do appreciate your guidance.

**Comment by tertsdiepraam** [comment]
Whoo, that's a lot of changes. Good job on keeping it all the tests passing😄 I think there is a lot of great stuff in here and I'll be a able to do a full review in a couple of days.

However, I'm probably going to request some changes, because I was not clear enough in my previous comment (sorry). The `OnceCell` was an optimization that I think we should keep, because it allows us to reduce syscalls and therefore makes `ls` faster [1]. Again, sorry for not being clear enough. I think we can still use most of this, though! Once I've read and understood it all I will try to work with you to merge this.

[1] With this optimization we don't ask for the metadata if we don't need it, if we do we fill the `OnceCell` and only retrieve it once. In this PR we always retrieve the metadata.

**Comment by kimono-koans** [comment]
> Whoo, that's a lot of changes. Good job on keeping it all the tests passing😄 I think there is a lot of great stuff in here and I'll be a able to do a full review in a couple of days.
> 
> However, I'm probably going to request some changes, because I was not clear enough in my previous comment (sorry). The `OnceCell` was an optimization that I think we should keep, because it allows us to reduce syscalls and therefore makes `ls` faster [1]. Again, sorry for not being clear enough. I think we can still use most of this, though! Once I've read and understood it all I will try to work with you to merge this.
> 
> [1] With this optimization we don't ask for the metadata if we don't need it, if we do we fill the `OnceCell` and only retrieve it once. In this PR we always retrieve the metadata.

Thank you for taking a look and for your continued review.

I pulled out OnceCell because I got sick of unwrapping everything twice.  And, although OnceCell *seems* like an optimization, I'm not sure it is.  Why not just request and store the metadata for every file?  We need that item metadata anyway to determine the file type, which we need to determine whether an item is a directory or a file, which we need to do for  *every item.*  I think the only reason for not storing the information is the cost of allocation?  My version is 4x faster on simple 'ls -R /' cases.  Right now, it's a lot of bureaucracy with much of an obvious win.

I know you wanted to return the error from any metadata request, which I think we can now do by simply changing the Option to a Result, Boxing the actual error, and sending that error back to whatever main logic we are executing, if we so choose.  As you know, I went back and forth on this(!).  However, it works fine now, and it seems to me that we would be allocating more per struct if we allocate for the Error type in the PathData struct, instead of a None?

Interested to know your thoughts!  Thanks.

**Comment by kimono-koans** [comment]
> Whoo, that's a lot of changes. Good job on keeping it all the tests passing😄 I think there is a lot of great stuff in here and I'll be a able to do a full review in a couple of days.
> 
> However, I'm probably going to request some changes, because I was not clear enough in my previous comment (sorry). The `OnceCell` was an optimization that I think we should keep, because it allows us to reduce syscalls and therefore makes `ls` faster [1]. Again, sorry for not being clear enough. I think we can still use most of this, though! Once I've read and understood it all I will try to work with you to merge this.
> 
> [1] With this optimization we don't ask for the metadata if we don't need it, if we do we fill the `OnceCell` and only retrieve it once. In this PR we always retrieve the metadata.

Of course, you're right -- the cost of allocating is a killer.  I'll withdraw this PR and re-submit it when I've worked my changes within the present OnceCell framework with the PathData struct.

Thank you for your review and I hope you will review again once I've cleaned it up!

**Pull request closed without merging**
